flopy/mf6/utils/codegen/dfn.py

```diff
@@ -0,0 +1,562 @@
+        """
+        Try to load a reference from the definition.
+        Returns `None` if the definition cannot be
+        referenced by other contexts.
+        """
+
+        # TODO: all this won't be necessary once we
+        # structure DFN format; we can then support
+        # subpackage references directly instead of
+        # by making assumptions about `dfn.meta`
+
+        if not dfn.meta or "dfn" not in dfn.meta:
+            return None
+
+        _, meta = dfn.meta["dfn"]
+
+        lines = {
+            "subpkg": next(
+                iter(
+                    m
+                    for m in meta
+                    if isinstance(m, str) and m.startswith("subpac")
+                ),
+                None,
+            ),
+            "parent": next(
+                iter(
+                    m
+                    for m in meta
+                    if isinstance(m, str) and m.startswith("parent")
+                ),
+                None,
+            ),
+        }
+
+        def _subpkg():
+            line = lines["subpkg"]
+            _, key, abbr, param, val = line.split()
+            matches = [v for v in dfn.values() if v["name"] == val]
+            if not any(matches):
+                descr = None
+            else:
+                if len(matches) > 1:
+                    warn(f"Multiple matches for referenced variable {val}")
+                match = matches[0]
+                descr = match["description"]
+
+            return {
+                "key": key,
+                "val": val,
+                "abbr": abbr,
+                "param": param,
+                "description": descr,
+            }
+
+        def _parent():
+            line = lines["parent"]
+            split = line.split()
+            return split[1]
+
+        return (
+            cls(**_subpkg(), parent=_parent())
+            if all(v for v in lines.values())
+            else None
+        )
+
+
+@renderable
+class Dfn(UserDict):
+    """
+    MODFLOW 6 input definition. An input definition
+    file specifies a component of an MF6 simulation,
+    e.g. a model or package.
+    """
+
+    class Name(NamedTuple):
+        """
+        Uniquely identifies an input definition.
+        Consists of a left term and a right term.
+        """
+
+        l: str
+        r: str
+
+        @classmethod
+        def parse(cls, v: str) -> "Dfn.Name":
+            try:
+                return cls(*v.split("-"))
+            except:
+                raise ValueError(f"Bad DFN name format: {v}")
+
+    name: Optional[Name]
+    meta: Optional[Dict[str, Any]]
+
+    def __init__(
+        self,
+        data: Optional[Vars] = None,
+        name: Optional[Name] = None,
+        meta: Optional[Dict[str, Any]] = None,
+    ):
+        self.data = OMD(data)
+        self.name = name
+        self.meta = meta
+
+    @staticmethod
+    def _load(f, common: Optional[dict] = None) -> Tuple[OMD, List[str]]:
```

Comment:
I was wondering if you would like to use that parser library that you found to set up a grammar, instead of writing all the push/pop logic yourself.

Reply:
imo it's not worth the effort since it's such a simple format and we'll be working with toml soon anyway, this can be thrown away soon. also lark was slower (admittedly I made no serious effort to optimize). if we use it to parse input files that will need attention